src/shared/condition.c

```diff
@@ -958,6 +960,118 @@ static int condition_test_file_is_executable(Condition *c, char **env) {
                 (st.st_mode & 0111));
 }
 
+static int condition_test_psi(Condition *c, char **env) {
+        _cleanup_free_ char *first = NULL, *second = NULL, *pressure_path = NULL;
+        loadavg_t *current, limit;
+        ResourcePressure pressure;
+        const char *p, *value;
+        OrderOperator order;
+        int r;
+
+        assert(c);
+        assert(c->parameter);
+        assert(IN_SET(c->type, CONDITION_MEMORY_PRESSURE, CONDITION_CPU_PRESSURE));
+
+        if (!is_pressure_supported())
+                return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Pressure Stall Information (PSI) is not supported");
+
+        p = c->parameter;
+        r = extract_many_words(&p, ":", 0, &first, &second, NULL);
+        if (r <= 0)
+                return log_debug_errno(r ?: SYNTHETIC_ERRNO(EINVAL), "Failed to parse condition parameter %s: %m", c->parameter);
+        /* If only one parameter is passed, then we look at the global system pressure rather than a specific cgroup. */
+        if (r == 1) {
+                pressure_path = path_join("/proc/pressure", c->type == CONDITION_CPU_PRESSURE ? "cpu" : "memory");
+                if (!pressure_path)
+                        return -ENOMEM;
+
+                value = first;
+        } else {
+                CGroupMask mask, required_mask = c->type == CONDITION_CPU_PRESSURE ? CGROUP_MASK_CPU : CGROUP_MASK_MEMORY;
+                const char *controller = c->type == CONDITION_CPU_PRESSURE ? "cpu.pressure" : "memory.pressure";
+                _cleanup_free_ char *slice_path = NULL;
+
+                r = cg_all_unified();
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to determine whether the unified cgroups hierarchy is used: %m");
+                if (r == 0)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Requires the unified cgroups hierarchy");
+
+                r = cg_mask_supported(&mask);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to get supported cgroup controllers: %m");
+
+                if (!FLAGS_SET(mask, required_mask))
+                        return log_debug_errno(SYNTHETIC_ERRNO(EOPNOTSUPP),
+                                               "Cgroup %s controller not available.",
+                                               required_mask == CGROUP_MASK_CPU ? "CPU" : "memory");
+
+                r = cg_slice_to_path(first, &slice_path);
+                if (r < 0)
+                        return log_debug_errno(r, "Cannot determine slice \"%s\" cgroup path: %m", first);
+
+                r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, slice_path, controller, &pressure_path);
```

Comment:
hmm, we should probably get PID1's own cgroup, then go one level up from that (i.e. drop the /init.scope/ part in the path). that way we'll look at the top of the current OS' cgroup tree.

Reply:
I don't follow - the intention here is to check an arbitrary slice

Reply:
the root of the cgroupfs and the root of what PID 1 manages are not necessarily the same. They often are, but we actually explicitly support if they are not. This has various usecases. Originally it was mostly about containers, i.e. when cgroup namespacing didn't exist: if we'd start a service manager in some cgroup down the tree it would take possesion of that and nothing else. It's still actively used for the --user stuff, there too we run things in a cgroup down the tree. But we officially allow people to also have some process as PID 1 before system starts that moves systemd down the cgroup tree somewhere, and leaves some sibling for other purposes outside of systemd's control.

Reply:
I get that, but I don't understand how to apply it here? It's already looking at the top of the OS tree - `cg_get_path` actually ignores the controller, and simply appends what you pass to `"/sys/fs/cgroup/`. So for example:

`user-1000.slice:100%` returns `/sys/fs/cgroup/user.slice/user-1000.slice/cpu.pressure`
`-.slice:100%` returns `/sys/fs/cgroup/cpu.pressure`
`bork.slice:100%` returns `/sys/fs/cgroup/bork.slice/cpu.pressure`

What's missing?

Reply:
@bluca I'd say the missing part is `cgroup_root` (such as in [unit_default_cgroup_path](https://github.com/systemd/systemd/blob/312dac28695ebe7c316c9b2dec40bc7182f5e28e/src/core/cgroup.c#L1934), alternatively [without the manager](https://github.com/systemd/systemd/blob/312dac28695ebe7c316c9b2dec40bc7182f5e28e/src/basic/cgroup-util.c#L2002)).

> But we officially allow people to also have some process as PID 1 before system starts that moves systemd down the cgroup tree somewhere, and leaves some sibling for other purposes outside of systemd's control.

@poettering Citation needed (I mean where's that documented?)

Reply:
how would the result be different? SYSTEMD_CGROUP_CONTROLLER is actually ignored since this only supports cgroups v2

Reply:
The result would be different for the user instance of systemd (or some non-standard tree layouts as implied by @poettering ). For the PID 1, there'd be indeed no difference.

Reply:
`cg_get_root_path` returns an empty path both under the system manager and the user instance

Reply:
Oops, I confused it with `manager_setup_cgroup` where it's not called with PID 1 but 0 (self) instead. The latter is what I meant would be missing in the user instance.
IOW, the condition in user instance should: a) without slice -- look at global pressure, b) with a slice specified -- look at user instance slice. Does that make sense?